egs/wsj/s5/steps/nnet3/convert_nnet2_to_nnet3.py

```diff
@@ -0,0 +1,441 @@
+        self.config = ""
+        self.transition_model = ""
+        self.components = []
+
+    def AddComponent(self, component, pairs):
+        '''
+        Adds components to the model given a 
+        dictionary of key-value pairs.
+        '''
+        self.components_read += 1
+
+        Component = namedtuple('Component', 'ident component pairs')
+
+        if "<InputDim>" in pairs and self.input_dim == -1:
+            self.input_dim = pairs["<InputDim>"]
+
+        # remove nnet2 specific tokens and catch descriptors
+        if component == '<PnormComponent>' and '<P>' in pairs:
+            pairs.pop('<P>')
+        elif component in SPLICE_COMPONENTS:
+            self.components.append(Component("splice", component, pairs))
+            return
+
+        # format pairs: {'<InputDim>':43} -> {'input-dim':43}
+        pairs = ['{0}={1}'.format(TokenToString(key),pairs[key]) for key in pairs]
+        
+        # keep track of layer type number (e.g. affine3)
+        node_name = NODE_NAMES[component]
+        self.counts[node_name] += 1
+
+        # e.g. affine3
+        ident = node_name + str(self.counts[node_name])
+
+        # <PnormComponent> -> PnormComponent
+        component = component[1:-1]
+
+        self.components.append(Component(ident, component, pairs))
+
+    def WriteConfig(self, filename):
+        '''
+        Write config to filename.
+        '''
+        logger.info('Writing config to {0}'.format(filename))
+
+        self.config = filename
+        with open(filename, 'w') as f:
+            for component in self.components:
+                if component.ident == "splice":
+                    continue
+                config_string = ' '.join(component.pairs)
+
+                f.write('component name={name} type={comp_type} {config_string}\n'.format(name=component.ident, comp_type=component.component, config_string=config_string))
+
+            f.write('\n# Component nodes\n')
+            f.write('input-node name=input dim={0}\n'.format(self.input_dim))
+            previous_component='input'
+            for component in self.components:
+                if component.ident == "splice":
+                    # Create splice string for the next node
+                    previous_component=MakeSpliceString(previous_component, component.pairs['<Context>'])
+                    continue
+                f.write('component-node name={name} component={name} input={inp}\n'.format(name=component.ident, inp=previous_component))
+                previous_component = component.ident
+            logger.warning('Assuming linear objective.')
+            f.write('output-node name=output input={inp} objective={obj}\n'.format(inp=previous_component, obj='linear'))
+
+    def WriteModel(self, model, binary="true"):
+        result = 0
+
+        # write raw model
+        result += subprocess.call('nnet3-init --binary=true {0} {1}'.format(self.config, os.path.join(tmpdir, 'nnet3.raw')), shell=True)
```

Comment:
Do you want to consider using the KaldiCommand function available in the nnet3 python libraries ? This might help you handle some common errors.